### ISSUE TITLE
WriteAsyncThreadSafe is not always safe

### ISSUE THREAD

**Issue by snakefoot** [Itoktsnhc/NLog.Azure#1]
Custom targets should only override this method if able to provide their own synchronization mechanism:

https://github.com/Itoktsnhc/NLog.Azure/blob/18c145cbc296bc2fcf7b68d8470335d57f6016a2/NLog.Azure/AzureTableTarget.cs#L48

The `Layout`-objects cannot always be guaranteed to be thread-safe, so using them inside `WriteAsyncThreadSafe` without using lock is dangerous.

Normally one should only override this method:

`protected virtual void Write(LogEventInfo logEvent)`

And if wanting to support batching, then this method:

`protected virtual void Write(IList<AsyncLogEventInfo> logEvents)`

**Comment by Itoktsnhc** [comment]
Hi snakefoot：
Thanks for reminding me of this wrong usage，I will fix it right away.
By the way, can you tell me more details about writing custom Nlog extensions? 
This link  https://github.com/nlog/nlog/wiki/How-to-write-a-custom-target is too general.
Many thanks

**Comment by Itoktsnhc** [comment]
@snakefoot  Because I am using AppendBlockAsync instead of AppendTextAsync. And AppendBlockAsync to an azure append blob is ok in muti-thread environment, this is why i cannot find the exception.
(Sorry for  my poor english)

**Comment by Itoktsnhc** [comment]
@snakefoot  I check my code under multi-threaded environment, the CloudQueueClient's AddMessageAsync、EventHubClient‘s SendAsync、CloudBlobClient's AppendBlockAsync and CloudTableClient ExecuteBatchAsync on different partitionkey and rowkey will be ok  on concurrent execution environment. So i think there should be no problem in this situation.

**Comment by Itoktsnhc** [comment]
I rechecked the code and added  lock to methods that use Layout.Render method.
https://github.com/Itoktsnhc/NLog.Azure/commit/98238583e079d7320aae7869faef7c68021575b8

**Comment by snakefoot** [comment]
@Itoktsnhc You can choose just to render under lock, and the rest without lock. So forexample it could be:

```
        protected override void WriteAsyncThreadSafe(IList<AsyncLogEventInfo> logEvents)
        {
             // Only acquire the _lockObject once for the entire payload, instead of for each object (very expensive)
            List<AzureQueueLogDto> payload;
            lock (_lockObject)
            {
                 payload = logEvents.Select(s => GenerateLogDto(s.LogEvent)).ToList();
            }
            WriteToQueueAsync(payload).GetAwaiter().GetResult();
        }

        protected override void WriteAsyncThreadSafe(AsyncLogEventInfo logEvent)
        {
            List<AzureQueueLogDto> payload;
            lock (_lockObject)
            {
                 payload = new List<AzureQueueLogDto>{ GenerateLogDto(logEvent.LogEvent) };
            }
            WriteToQueueAsync(payload).GetAwaiter().GetResult();
        }
```

**Comment by snakefoot** [comment]
> By the way, can you tell me more details about writing custom Nlog extensions?

Right now it is a little difficult to write custom NLog targets, when wanting to support batching and flow-control along with capture of async state. We hope to make this better in the future, so it is possible to write high performance custom NLog targets without stepping on too many bombs.

Can give you one tip, and that is the property `Target.OptimizeBufferReuse = true` and `Target.RenderLogEvent`. `Target.RenderLogEvent` can be used instead of `Layout.Render` and will reduce string-allocations (Must be used under lock, like `Layout.Render`)

**Comment by Itoktsnhc** [comment]
@snakefoot Thanks for the help

**Comment by snakefoot** [comment]
Fixed with 71524b8beb1ce1f0e4d0eb70a2f3786ccb770217
